### ISSUE TITLE
GLM: Allow scalar, row vector for both lpmdfs/lpdfs

### ISSUE THREAD

**Issue by drezap** [stan-dev/math#1269]
## Description
This is regarding @avehtari https://github.com/stan-dev/math/pull/1242#issuecomment-496652705.

More generally, I think we have an issue with type casting in the stan language. It's very inconvenient for GLMs to take in arbitrary datatypes (for example using a vector instead of having to cast into an Nx1 array for inverse logit). 

@seantalts  Not up to date on stanc3, but if this will handle this, then we can close this issue.

Any reasons why we shouldn't support this?

Could drastically increase the amount of tests.

Might be easy to write a meta program that automatically casts these data types into the format that the pdfs/pmfs take in, to cut down on the amount of code we have to write. @syclik thoughts on this?

## Example
GLMs to take in more types than the rigid set they have now.

## Expected Output
Compiling Stan models without unnecessary casting.

#### Current Version:
v2.19.1

**Comment by t4c1** [comment]
> Might be easy to write a meta program that automatically casts these data types into the format that the pdfs/pmfs take in, to cut down on the amount of code we have to write, and testing we have to do.

All existing GLMs are implemented more or less this way. Any vector argument can be of type `vector`, `row_vector` or `array`. However I don't think all input combinations are tested or exposed to stan language. So this will likely only involve adding a few tests and exposing the rest of interfaces to stan language.

**Comment by drezap** [comment]
updated comments according to t4c1's input.

**Comment by drezap** [comment]
Some of these aren't templated to allow all of these input types, making a list as I go through:

categorical
categorical_logit
ordered_probit
and many more...

Yeah, unfortunately many are just taking Eigen column vectors in, so I'd have to go through and re-template all of our GLMs, which would be a pain.

Let me see if I can think of way to do it fast with a script without having to do each one by hand. I'm going to leave this alone and come back to it.

**Comment by drezap** [comment]
Instead of re-templating everything that only takes in Eigen column vectors, would anyone be opposed to me just writing a function that casts vectors, scalars, row_vectors into Eigen column vectors, and just throwing exception handling that catches the ones that are not Column vectors and calls a function that casts them? Is this sloppy? Is there a better way?

**Comment by t4c1** [comment]
> Some of these aren't templated to allow all of these input types

ONly GLMs are written this way, not every distribution. That is `normal_id_glm_lpdf`, `bernoulli_beta_glm_lpmf`, `poisson_log_glm_lpmf` and `negbin_2_log_glm_lpmf`.

There already is a function that creates a column vector view from these types - `as_column_vector_or_scalar()`. However unless all these functions are changed to accept these views there will be some unnecessary copying of data done if any of these functions are called with that view as an argument. 

One good option is to template all these functions. Another, in my opinion a better option is to make all stan functions accept arbitrary EIgen expressions. However that would be even more work.

**Comment by bob-carpenter** [comment]
> On Jun 20, 2019, at 2:05 AM, t4c1 <notifications@github.com> wrote:
> 
> Some of these aren't templated to allow all of these input types
> 
> ONly GLMs are written this way, not every distribution. That is normal_id_glm_lpdf, bernoulli_beta_glm_lpmf, poisson_log_glm_lpmf and negbin_2_log_glm_lpmf.
> 
> There already is a function that creates a column vector view from these types - as_column_vector_or_scalar(). However unless all these functions are changed to accept these views there will be some unnecessary copying of data done if any of these functions are called with that view as an argument.

Right.  You can look at how this is handled in the lpdf and lpmf definitions.
All inputs are wrapped in views that automatically broadcast (repeat
scalar inputs to match vector input sizes) if necessary.


> One good option is to template all these functions. Another, in my opinion a better option is to make all stan functions accept arbitrary EIgen expressions. However that would be even more work.

That's the long term goal, but I don't think it's quite possible as they have two distinct classes of expressions.  But we could get most of the way there.

**Comment by t4c1** [comment]
> two distinct classes of expressions

I did not know about this. Which two?

**Comment by bob-carpenter** [comment]
We've had a lot of trouble in the past consolidating standard arithmetic templates, like you get from `operator+` or `transpose` with those you get from `block`.  I've always found [Eigen's doc on writing functions](https://eigen.tuxfamily.org/dox/TopicFunctionTakingEigenTypes.html) to be very confusing because it's very tied up in their use of CRTP in their low-level implementations.

But you can see right from the doc that their base CRTP classes like `DenseBase` don't quite play nicely with everyting;  for instance, `DenseBase` says "Accepts any dense matrix or array expression, but no sparse objects and no special matrix classes such as DiagonalMatrix."  Then I'm never sure how wide a class of objects can be used with `Ref`.  I can't recall if `MatrixBase`, which they recommend in the summary, has similar problems.

**Comment by drezap** [comment]
Is there any tests for bernoulli_beta_glm_lpmf? I'm not finding them.

**Comment by t4c1** [comment]
they are in `test/unit/math/rev/mat/prob`

**Comment by drezap** [comment]
Yeah, there are no tests for bernoulli_beta_glm_lpmf. Nor is there a bernoulli_beta_glm_lpmf.hpp. Unless you mean bernoulli_logit? I've done this one instead. I know beta is conjugate prior to bernoulli but I'm not sure why there'd be a specialized pmf.

Also, doesn't GLM refer to any linear model with a non-gaussian likelihood that requires some transformation to constrain values to a particular range?

Anyway. I've done the 4, and I'll open up an issue where I'll start templating all of our pmfs/pdfs to make our language more flexible. This will cover everything.

**Comment by t4c1** [comment]
Yeah, I meant bernoulli_logit_glm_lpmf. Though there is afaik no bernoulli_beta_glm_lpmf in stan.

**Comment by bob-carpenter** [comment]
Right.  The naming convention likelihood_link_glm_lp{d|m}f.  I don't think a beta-bernoulli log beta-binomial can be formulated as GLMs. They're both conjugate, so the ideal implementation would be as described in the user's guide chapter on efficiency (in short, code the log posterior rather than sum of log likelihood and log prior).

**Comment by SteveBronder** [comment]
Superseded by https://github.com/stan-dev/math/issues/1378 and issue was closed over there
